Reduce unsafe buffer usage in WebCore::GraphicsContextGLANGLE

geoffreygaren opened the pull request:
#### c66cda75a86abab2961a635b504b297e85c8a3c9
<pre>
Reduce unsafe buffer usage in WebCore::GraphicsContextGLANGLE
<a href="https://bugs.webkit.org/show_bug.cgi?id=286036">https://bugs.webkit.org/show_bug.cgi?id=286036</a>
<a href="https://rdar.apple.com/143004066">rdar://143004066</a>

Reviewed by Chris Dumez.

* Source/WebCore/platform/graphics/GCGLSpan.h:
(GCGLSpanTuple::span const): Added a span accessor so clients can be bounds safe.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(asPointers): Helper function to avoid duplicate code. Uses span now.

(WebCore::GraphicsContextGLANGLE::multiDrawElementsANGLE):
(WebCore::GraphicsContextGLANGLE::multiDrawElementsInstancedANGLE):
(WebCore::GraphicsContextGLANGLE::multiDrawElementsInstancedBaseVertexBaseInstanceANGLE):
Use our new bounds-safe helper function.

Canonical link: <a href="https://commits.webkit.org/288987@main">https://commits.webkit.org/288987@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/53743e951bf020cc2db286f50427449b49b94f07

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84942 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4667 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90091 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35999 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12644 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66105 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23921 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3639 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77195 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46372 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3513 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31419 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35071 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32227 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91469 "Hash 53743e95 for PR 39105 does not build (failure)") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12281 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/91469 "Hash 53743e95 for PR 39105 does not build (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12511 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73003 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/91469 "Hash 53743e95 for PR 39105 does not build (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18093 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16543 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13242 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12233 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/17686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12068 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15562 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13813 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -73,6 +73,17 @@ static void wipeAlphaChannelFromPixels(std::span<uint8_t> pixels)
+    // Must cast offsets from int to void* before passing down to ANGLE.
+
+    Vector<const void*> pointers;
+    pointers.reserveInitialCapacity(offsets.size());
```
Ideally, this would be using `WTF::map()`, which is potentially more efficient because it avoids the capacity checks in `append()`.

geoffreygaren replied:
Less code too. I can do that.

geoffreygaren committed:
Reduce unsafe buffer usage in WebCore::GraphicsContextGLANGLE
https://bugs.webkit.org/show_bug.cgi?id=286036
rdar://143004066

Reviewed by Chris Dumez.

* Source/WebCore/platform/graphics/GCGLSpan.h:
(GCGLSpanTuple::span const): Added a span accessor so clients can be bounds safe.

* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(asPointers): Helper function to avoid duplicate code. Uses span now.

(WebCore::GraphicsContextGLANGLE::multiDrawElementsANGLE):
(WebCore::GraphicsContextGLANGLE::multiDrawElementsInstancedANGLE):
(WebCore::GraphicsContextGLANGLE::multiDrawElementsInstancedBaseVertexBaseInstanceANGLE):
Use our new bounds-safe helper function.

Canonical link: https://commits.webkit.org/288987@main

darinadler commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -73,6 +73,15 @@ static void wipeAlphaChannelFromPixels(std::span<uint8_t> pixels)
 }
 #endif
 
+static inline const Vector<const void*> asPointers(std::span<const GCGLsizei> offsets)
```
The `const` here in `const Vector` is not helpful and should probably be removed

Also for code like this we can also consider returning `FixedVector` for slightly better efficiency, if we had a `WTF::map` that supported; no particular reason to prefer `Vector` with its slightly larger size and capability to be efficiently resized after creation
